token-metadata/js/src/instruction.ts

```diff
@@ -0,0 +1,53 @@
+import type { PublicKey } from '@solana/web3.js';
+
+/** The field can be one of the required fields (name, symbol, URI), or a
+ * totally new field denoted by a "key" string. */
+type Field = 'name' | 'symbol' | 'uri' | string;
+
+export interface TokenMetadata {
+    // The authority that can sign to update the metadata
+    updateAuthority: PublicKey;
+    // The associated mint, used to counter spoofing to be sure that metadata belongs to a particular mint
+    mint: PublicKey;
+    // The longer name of the token
+    name: string;
+    // The shortened symbol for the token
+    symbol: string;
+    /// The URI pointing to richer metadata
+    uri: string;
+    /// Any additional metadata about the token as key-value pairs
+    additionalMetadata: [string, string][];
+}
+
+export interface TokenMetadataInstruction {
+    /** If the field does not exist on the account, it will be created.
+     * If the field does exist, it will be overwritten. */
+    updateField: (
+        programId: PublicKey,
+        metadata: PublicKey,
+        updateAuthority: PublicKey,
+        field: Field,
+        value: string
+    ) => Promise<void>;
+
+    /** Removes a key-value pair in a token-metadata account. This only applies
+     * to additional fields, and not the base name / symbol / URI fields. */
+    removeKey: (
+        programId: PublicKey,
+        metadata: PublicKey,
+        updateAuthority: PublicKey,
+        field: Field,
+        idempotent: boolean
+    ) => Promise<void>;
+
+    /** Updates the token-metadata authority */
+    updateAuthority: (
+        programId: PublicKey,
+        metadata: PublicKey,
+        oldAuthority: PublicKey,
+        newAuthority: PublicKey
+    ) => Promise<void>;
+
+    // Emits the token-metadata as return data
+    emit: (programId: PublicKey, metadata: PublicKey) => Promise<TokenMetadata>;
+}
```

Comment:
I don't think this actually needs to be an interface itself. I think maybe we should just define a set of functions for creating the interface instructions, similar to the functions that are laid out here:

https://github.com/solana-labs/solana-program-library/blob/d1f06df779bbf52eb7df85300b619ed6d6675776/token-metadata/interface/src/instruction.rs#L225-L319

Think about it like this: The interface itself is for on-chain programs, so the JS library doesn't _itself_ have to be an actual interface. Rather, it's a library designed for working with programs that implement the _Rust interface_.

Lmk if this makes sense!

Reply:
You are also missing one: `initialize`.

And we'll need to roll some `pack/unpack` stuff as well using TLV or manually packing the bytes (likely the latter).

Reply:
Makes sense - functions provides more flexibility for sure. 

I've added `initialize`, but unsure what the return type should be - got any input? Should it be done as a generic? 

I've added the `pack/unpack` interfaces, still need to figure out the implementation so will add later

Reply:
Left comments in the code on the types/functions you laid out